File: app/src/main/java/com/bachors/iptv/PlaylistActivity.java
package com.bachors.iptv;

import android.annotation.SuppressLint;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;

import com.bachors.iptv.adapter.PlaylistAdapter;
import com.bachors.iptv.models.PlaylistData;
import com.bachors.iptv.models.PlaylistResponse;
import com.bachors.iptv.util.HttpHandler;
import com.bachors.iptv.util.RecyclerTouchListener;
import com.bachors.iptv.util.SharedPrefManager;
import com.google.gson.Gson;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class PlaylistActivity extends AppCompatActivity {

    private static final String EXT_M3U = "#EXTM3U";
    private static final String EXT_INF = "#EXTINF:";
    private static final String EXT_LOGO = "tvg-logo";
    private static final String EXT_URL = "http://";

    String stream;
    String goLink;
    String goTitle;
    String goJson;

    ProgressDialog loading;

    SharedPrefManager sharedPrefManager;
    int key;
    int index = 0;

    Context mcon;

    List<PlaylistData> allData = new ArrayList<>();
    List<PlaylistData> searchData;
    boolean all = true;
    PlaylistAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_playlist);

        mcon = this;

        sharedPrefManager = new SharedPrefManager(this);

        adapter = new PlaylistAdapter(this);

        Spinner spinner = (Spinner) findViewById(R.id.playlist);
        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                index = position;
                adapter.clear();
                jsonTogson();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });
        List<String> pl = new ArrayList<String>();
        pl.add("Category");
        pl.add("Language");
        pl.add("Country");
        ArrayAdapter<String> dataAdapter = new ArrayAdapter<String>(this, android.R.layout.simple_spinner_item, pl);
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(dataAdapter);

        RecyclerView rv = (RecyclerView) findViewById(R.id.rv);
        GridLayoutManager gl = new GridLayoutManager(this, 2);
        rv.setLayoutManager(gl);
        rv.setItemAnimator(new DefaultItemAnimator());
        rv.setAdapter(adapter);
        rv.addOnItemTouchListener(new RecyclerTouchListener(this, rv, new RecyclerTouchListener.ClickListener() {
            @Override
            public void onClick(View view, int position) {
                goTo(position);
            }

            @Override
            public void onLongClick(View view, int position) {

            }
        }));
        adapter.clear();
        jsonTogson();

    }

    private void jsonTogson() {
        try {
            JSONArray ar = new JSONArray(sharedPrefManager.getSpPlaylist());
            JSONObject ob = new JSONObject();
            ob.put("data", ar.getJSONArray(index));
            Gson gson = new Gson();
            PlaylistResponse playlist = gson.fromJson(ob.toString(), PlaylistResponse.class);
            allData = playlist.getData();
            adapter.addAll(allData);
        } catch (JSONException e) {
            e.printStackTrace();
        }

    }

    private void filter(String text) {
        //new array list that will hold the filtered data
        searchData = new ArrayList<>();

        //looping through existing elements
        for (PlaylistData s : allData) {
            //if the existing elements contains the search input
            if (s.getTitle().toLowerCase().contains(text.toLowerCase())) {
                //adding the element to filtered list
                searchData.add(s);
            }
        }

        //calling a method of the adapter class and passing the filtered list
        adapter.addAll(searchData);
    }

    private void goTo(int id){
        if(all){
            key = id;
        }else{
            key = allData.indexOf(searchData.get(id));
        }

        goTitle = allData.get(key).getTitle();
        goLink = allData.get(key).getLink();

        loading = ProgressDialog.show(mcon, null, "Loading ...", true, false);
        new m3uParse().execute();

    }

    @Override
    public void onBackPressed() {
        finish();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        getMenuInflater().inflate(R.menu.menu_search, menu);
        getMenuInflater().inflate(R.menu.menu_load, menu);

        MenuItem mSearch = menu.findItem(R.id.action_search);

        SearchView mSearchView = (SearchView) mSearch.getActionView();
        mSearchView.setQueryHint("Search...");

        mSearchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                adapter.clear();
                //after the change calling the method and passing the search input
                if(s.equals("")){
                    all = true;
                    adapter.addAll(allData);
                }else {
                    all = false;
                    filter(s);
                }
                return true;
            }
        });

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        MenuItem actionViewItem2 = menu.findItem(R.id.load);
        // Retrieve the action-view from menu
        View v = MenuItemCompat.getActionView(actionViewItem2);
        Button x = (Button) v.findViewById(R.id.btn_load);
        // Handle button click here
        x.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loading = ProgressDialog.show(mcon, null, "Update data ...", true, false);
                loadMain();
            }
        });
        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_about) {
            LayoutInflater inflater = this.getLayoutInflater();
            View dialogView = inflater.inflate(R.layout.bachors_apps, null);
            new AlertDialog.Builder(this)
                    .setCancelable(false)
                    .setNegativeButton("Ok", null)
                    .setView(dialogView)
                    .show();
        }else if (id == R.id.action_rate) {
            Uri uri = Uri.parse("https://github.com/bachors/IPTV-Android");
            Intent intent = new Intent(Intent.ACTION_VIEW, uri);
            startActivity(intent);
        }else if (id == R.id.action_share) {
            Intent shareIntent = new Intent(Intent.ACTION_SEND);
            shareIntent.setType("text/plain");
            shareIntent.putExtra(Intent.EXTRA_TEXT, "https://github.com/bachors/IPTV-Android");
            startActivity(Intent.createChooser(shareIntent, "Share link using"));
        }

        return super.onOptionsItemSelected(item);
    }

    private void loadMain() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                final JSONArray json = new JSONArray();

                try {
                    Document doc = Jsoup.connect("https://raw.githubusercontent.com/iptv-org/iptv/master/README.md").get();

                    Elements tables = doc.select("table");

                    for (Element table : tables) {
                        JSONArray ar = new JSONArray();
                        Elements trs = table.select("tr");
                        int i = 0;
                        for (Element tr : trs) {
                            if(i > 0) {
                                JSONObject ob = new JSONObject();
                                ob.put("title", tr.select("td").eq(0).text());
                                ob.put("link", tr.select("td").eq(2).select("code").eq(0).text());
                                ar.put(ob);
                            }
                            i++;
                        }
                        json.put( ar);
                    }
                } catch (IOException ignored) {

                } catch (JSONException e) {
                    e.printStackTrace();
                }

                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        loading.dismiss();

                        sharedPrefManager.saveSPString(SharedPrefManager.SP_PLAYLIST, json.toString());

                        Intent launchNextActivity;
                        launchNextActivity = new Intent(getApplicationContext(), PlaylistActivity.class);
                        launchNextActivity.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                        launchNextActivity.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK);
                        launchNextActivity.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
                        startActivity(launchNextActivity);
                        overridePendingTransition(0, 0);
                    }
                });
            }
        }).start();
    }

    /**
     * Async task class to get json by making HTTP call
     */
    @SuppressLint("StaticFieldLeak")
    private class m3uParse extends AsyncTask<Void, Void, Void> {

        @Override
        protected void onPreExecute() {
            super.onPreExecute();

        }

        @Override
        protected Void doInBackground(Void... arg0) {
            HttpHandler sh = new HttpHandler();

            // Making a request to url and getting response
            String result = sh.makeServiceCall(goLink);

            if (result != null) {
                stream = result;
            }

            return null;
        }

        @Override
        protected void onPostExecute(Void result) {
            super.onPostExecute(result);
            loading.dismiss();

            String[] linesArray = stream.split(EXT_INF);
                JSONArray ar = new JSONArray();
                for (String currLine : linesArray) {
                    if (!currLine.contains(EXT_M3U)) {
                        JSONObject ob = new JSONObject();
                        String[] dataArray = currLine.split(",");
                        try {
                            String name = dataArray[1].substring(0, dataArray[1].indexOf(EXT_URL)).replace("\n", "");
                            String url = dataArray[1].substring(dataArray[1].indexOf(EXT_URL)).replace("\n", "").replace("\r", "");
                            ob.put("name", name);
                            ob.put("url", url);
                            if (dataArray[0].contains(EXT_LOGO)) {
                                String logo = dataArray[0].substring(dataArray[0].indexOf(EXT_LOGO) + EXT_LOGO.length()).replace("=", "").replace("\"", "").replace("\n", "");
                                ob.put("logo", logo);
                            } else {
                                ob.put("logo", "");
                            }

                            ar.put(ob);
                        } catch (Exception fdfd) {
                            Log.e("Google", "Error: " + fdfd.fillInStackTrace());
                        }
                    }
                }

                goJson = ar.toString();

                sharedPrefManager.saveSPString(SharedPrefManager.SP_CHANNELS, goJson);
                Intent intent = new Intent(mcon, ChannelsActivity.class);
                intent.putExtra("title", allData.get(key).getTitle());
                startActivity(intent);
        }

    }

}

File: app/src/main/java/com/bachors/iptv/PlaylistActivity.kt
package com.bachors.iptv

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ArrayAdapter
import android.widget.ImageView
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.core.graphics.drawable.toDrawable
import androidx.core.net.toUri
import androidx.core.view.WindowInsetsControllerCompat
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bachors.iptv.adapters.PlaylistAdapter
import com.bachors.iptv.databinding.ActivityPlaylistBinding
import com.bachors.iptv.models.PlaylistData
import com.bachors.iptv.utils.HttpHandler
import com.bachors.iptv.utils.RecyclerTouchListener
import com.bachors.iptv.utils.SharedPrefManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.json.JSONArray
import org.json.JSONObject
import org.jsoup.Jsoup

class PlaylistActivity : AppCompatActivity() {
    companion object {
        private const val EXT_M3U = "#EXTM3U"
        private const val EXT_INF = "#EXTINF:"
        private const val EXT_LOGO = "tvg-logo"
        private const val EXT_HTTP = "http://"
        private const val EXT_HTTPS = "https://"
    }

    private var stream: String? = null
    private var goLink: String? = null
    private var goTitle: String? = null
    private var goJson: String? = null
    private lateinit var loading: AlertDialog
    private lateinit var sharedPrefManager: SharedPrefManager
    private var key: Int = 0
    private var index: Int = 0
    private lateinit var mcon: Context
    private val allData = mutableListOf<PlaylistData>()
    private var searchData: List<PlaylistData>? = null
    private var all = true
    private lateinit var adapter: PlaylistAdapter
    var binding: ActivityPlaylistBinding? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityPlaylistBinding.inflate(layoutInflater)
        setContentView(binding?.root)
        supportActionBar?.elevation = 0f
        supportActionBar?.setBackgroundDrawable(Color.TRANSPARENT.toDrawable())
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        mcon = this

        val decorView = window.decorView
        val wic = WindowInsetsControllerCompat(window, decorView)
        wic.isAppearanceLightStatusBars = false
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            decorView.isForceDarkAllowed = true
        }

        sharedPrefManager = SharedPrefManager(this)

        val builder2 = MaterialAlertDialogBuilder(this, R.style.MyDialogTheme)
        builder2.setCancelable(false)
        builder2.setMessage("Please wait...")
        loading = builder2.create()

        adapter = PlaylistAdapter(this)

        setupAk()
        binding?.playlist?.setOnItemClickListener({ parent, view, position, id ->
            index = position
            adapter.clear()
            jsonTogson()
        })

        val linearLayoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        val rv = findViewById<RecyclerView>(R.id.rv)
        rv.layoutManager = linearLayoutManager
        rv.itemAnimator = DefaultItemAnimator()
        rv.adapter = adapter
        rv.addOnItemTouchListener(RecyclerTouchListener(this, rv, object : RecyclerTouchListener.ClickListener {
            override fun onClick(view: View, position: Int) {
                goTo(position)
            }

            override fun onLongClick(view: View, position: Int) {}
        }))

        val refresh = findViewById<ImageView>(R.id.btn_load)
        refresh.setOnClickListener {
            loading.show()
            loadPlaylists()
        }

        if (sharedPrefManager.getSpPlaylist().isEmpty() || sharedPrefManager.getSpPlaylist() == "[]") {
            loading.show()
            loadPlaylists()
        } else {
            adapter.clear()
            jsonTogson()
        }

        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                finish()
            }
        })
    }

    fun setupAk() {
        val pl = mutableListOf<String>()
        pl.add("Category")
        pl.add("Language")
        val dataAdapter = ArrayAdapter(this, android.R.layout.simple_dropdown_item_1line, pl)
        binding?.playlist?.setAdapter(dataAdapter);
        binding?.playlist?.setText(pl[0], false)
    }

    private fun jsonTogson() {
        try {
            val ar = JSONArray(sharedPrefManager.getSpPlaylist())
            val listType = object : TypeToken<List<PlaylistData>>() {}.type
            val gson = Gson()
            val data: List<PlaylistData> = gson.fromJson(ar.getJSONArray(index).toString(), listType)
            allData.clear()
            allData.addAll(data)
            adapter.addAll(allData)
        } catch (_: Exception) {
            // Handle exception
        }
    }

    private fun filter(text: String) {
        searchData = allData.filter { it.title.lowercase().contains(text.lowercase()) }
        adapter.addAll(searchData!!)
    }

    private fun goTo(id: Int) {
        key = if (all) id else allData.indexOf(searchData!![id])
        goTitle = allData[key].title
        goLink = allData[key].link
        loading.show()
        loadChannels()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_search, menu)
        menuInflater.inflate(R.menu.menu_download, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val searchView = menu.findItem(R.id.action_search).actionView as SearchView
        searchView.queryHint = "Search..."
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean = false

            override fun onQueryTextChange(s: String): Boolean {
                adapter.clear()
                if (s.isEmpty()) {
                    all = true
                    adapter.addAll(allData)
                } else {
                    all = false
                    filter(s)
                }
                return true
            }
        })
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        if (item.itemId == R.id.download) {
            val uri = "https://github.com/bachors/IPTV-Android".toUri()
            startActivity(Intent(Intent.ACTION_VIEW, uri))
        }
        return super.onOptionsItemSelected(item)
    }

    private fun loadPlaylists() {
        Thread {
            val json = JSONArray()
            try {
                val doc = Jsoup.connect("https://raw.githubusercontent.com/iptv-org/iptv/master/PLAYLISTS.md").get()
                val tables = doc.select("table")
                for (table in tables) {
                    val ar = JSONArray()
                    val trs = table.select("tr")
                    var i = 0
                    for (tr in trs) {
                        if (i > 0) {
                            val ob = JSONObject()
                            ob.put("title", tr.select("td").eq(0).text())
                            ob.put("channel", tr.select("td").eq(1).text())
                            ob.put("link", tr.select("td").eq(2).select("code").eq(0).text())
                            ar.put(ob)
                        }
                        i++
                    }
                    json.put(ar)
                }
            } catch (_: Exception) {}

            runOnUiThread {
                loading.dismiss()
                sharedPrefManager.saveSPString(SharedPrefManager.SP_PLAYLIST, json.toString())
                adapter.clear()
                jsonTogson()
            }
        }.start()
    }

    private fun loadChannels() {
        Thread {
            val sh = HttpHandler()
            val result = sh.makeServiceCall(goLink)
            if (result != null) {
                stream = result.replace("#EXTVLCOPT:(.*)".toRegex(), "")
            }

            runOnUiThread {
                loading.dismiss()
                val linesArray = stream!!.split(EXT_INF)
                val ar = JSONArray()
                for (currLine in linesArray) {
                    if (!currLine.contains(EXT_M3U)) {
                        val ob = JSONObject()
                        val dataArray = currLine.split(",")
                        try {
                            val name: String
                            val url: String
                            if (dataArray[1].contains(EXT_HTTPS)) {
                                name = dataArray[1].substring(0, dataArray[1].indexOf(EXT_HTTPS)).replace("\n", "")
                                url = dataArray[1].substring(dataArray[1].indexOf(EXT_HTTPS)).replace("\n", "").replace("\r", "")
                            } else {
                                name = dataArray[1].substring(0, dataArray[1].indexOf(EXT_HTTP)).replace("\n", "")
                                url = dataArray[1].substring(dataArray[1].indexOf(EXT_HTTP)).replace("\n", "").replace("\r", "")
                            }
                            ob.put("name", name)
                            ob.put("url", url)
                            if (dataArray[0].contains(EXT_LOGO)) {
                                val logo = dataArray[0].substring(dataArray[0].indexOf(EXT_LOGO) + EXT_LOGO.length).replace("=", "").replace("\"", "").replace("\n", "")
                                ob.put("logo", logo)
                            } else {
                                ob.put("logo", "")
                            }
                            ar.put(ob)
                        } catch (fdfd: Exception) {
                            Log.e("Google", "Error: ${fdfd.fillInStackTrace()}")
                        }
                    }
                }

                goJson = ar.toString()
                sharedPrefManager.saveSPString(SharedPrefManager.SP_CHANNELS, goJson!!)
                val intent = Intent(mcon, ChannelsActivity::class.java)
                intent.putExtra("title", allData[key].title)
                startActivity(intent)
            }
        }.start()
    }

    override fun onResume() {
        super.onResume()
        setupAk()
    }
}